### PR TITLE
[DeviceSanitizer] Support "-fsanitize-ignorelist=" to disable sanitizing on some of kernels

### PR DESCRIPTION
**Pull request by AllanZyne**
LLVM: https://github.com/intel/llvm/pull/15294

Add a device global "__AsanKernelMetadata" which records the sanitized kernel name, so that we can check if sanitizer layer needs specially handle it.

### CODE REVIEW

**Comment by pbalcer** [comment]
please rebase

**Review by zhaomaosu — APPROVED** [review]
lgtm

**Comment by AllanZyne** [comment]
Hi @oneapi-src/unified-runtime-maintain, please review, thank you.

File: source/loader/layers/sanitizer/asan/asan_interceptor.hpp [hunk 1]
```diff
@@ -266,8 +272,10 @@ class AsanInterceptor {
 
     std::shared_ptr<KernelInfo> getKernelInfo(ur_kernel_handle_t Kernel) {
         std::shared_lock<ur_shared_mutex> Guard(m_KernelMapMutex);
-        assert(m_KernelMap.find(Kernel) != m_KernelMap.end());
-        return m_KernelMap[Kernel];
+        if (m_KernelMap.find(Kernel) != m_KernelMap.end()) {
+            return m_KernelMap[Kernel];
+        }
+        return nullptr;
```
[the remark this answers is not in the document]
Comment (yingcong-wu) [reply source/loader/layers/sanitizer/asan/asan_interceptor.hpp:278]: Actually, I think this is better(return nullptr when not found), then there will be a segv that we will not miss. But adding assertion after a call that may return nullptr is better.

[the remark this answers is not in the document]
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan/asan_interceptor.hpp:278]: It needn't assert, I did this on purpose. 
It possible that some of kernel are not in `m_KernelMap`.
I have added "if" checkers when calling `getKernelInfo`.

[the remark this answers is not in the document]
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan/asan_interceptor.hpp:278]: KernelInfo doesn't like other XXXinfo. 
It's possible that some of kernel are not in m_KernelMap.

File: source/loader/layers/sanitizer/asan/asan_interceptor.hpp [hunk 2]
```diff
@@ -111,9 +111,9 @@ struct ProgramInfo {
     ur_program_handle_t Handle;
     std::atomic<int32_t> RefCount = 1;
 
-    // lock this mutex if following fields are accessed
```
[the remark this answers is not in the document]
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan/asan_interceptor.hpp:114]: Synced with @zhaomaosu, mutex can be removed here.
I removed "likely" in comment.

**Review by yingcong-wu — APPROVED** [review]

File: source/loader/layers/sanitizer/asan/asan_interceptor.cpp [hunk 3]
```diff
@@ -412,10 +413,106 @@ AsanInterceptor::updateShadowMemory(std::shared_ptr<ContextInfo> &ContextInfo,
...
+        Result = registerSpirKernels(Program);
+        if (Result != UR_RESULT_SUCCESS) {
+            break;
+        }
+        getContext()->logger.info("registerDeviceGlobals");
+        Result = registerDeviceGlobals(Program);
+        if (Result != UR_RESULT_SUCCESS) {
+            break;
+        }
+    } while (false);
```
Comment (pbalcer) [inline source/loader/layers/sanitizer/asan/asan_interceptor.cpp:429]: why a loop?

**Review by pbalcer — APPROVED** [review]
lgtm, just remove the unnecessary loop.

File: source/loader/layers/sanitizer/asan/asan_interceptor.cpp (hunk 3 above)
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan/asan_interceptor.cpp]: Removed

**Pull request merged**